### PR TITLE
FirstCommit(FirstUpdate)

### PR DESCRIPTION
**Pull request by mkelshapik**

### CODE REVIEW

File: src/MyArrayList.java [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+        } else {
+            throw new IllegalArgumentException("Неправильно вказаний індекс");
+        }
+    }
+
+    @Override
+    public boolean add(Object o) {
+        if (size >= capacity) {
+            T[] temp = t;
+            t = (T[]) new Object[temp.length + INITIAL_CAPACITY];
```
Comment (r-zaiarnyi) [inline src/MyArrayList.java:34]: you increased array but forget to increase capacity

File: src/MyArrayList.java [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+        if (index > size || index < 0) throw new IllegalArgumentException("Неправильно вказаний індекс");
+        if (index == 0){
+            T[] temp = t;
+            t = (T[]) new Object[temp.length + 1];
+            System.arraycopy(temp, 0, t, 1, temp.length);
+            t[index] = (T) element;
+        } else if (index == size){
+            if (capacity == size){
+                T[] temp = t;
+                t = (T[]) new Object[temp.length + INITIAL_CAPACITY];
```
Comment (r-zaiarnyi) [inline src/MyArrayList.java:53]: same as previous

File: src/MyArrayList.java [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    @Override
+    public Object remove(int index) {
+        if (index >= size || index < 0) throw new IllegalArgumentException("Неправильно вказаний індекс");
+        if (index == t.length - 1){
+            T[] temp = t;
+            t = (T[]) new Object[size - 1];
+            System.arraycopy(temp, 0, t, 0, t.length);
+            size--;
```
Comment (r-zaiarnyi) [inline src/MyArrayList.java]: you can do this in one place

File: src/MyArrayList.java [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    @Override
+    public int size() {
+        return size;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        if (size == 0) {
```
Comment (r-zaiarnyi) [inline src/MyArrayList.java]: you can just `return size == 0;`

File: src/MyArrayList.java [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+
+    @Override
+    public Object get(int index) {
+        return t[index];
+    }
+
+    @Override
+    public boolean contains(Object o) {
+        for (int i = 0; i < size; i++){
+            if (t[i].equals(o) && t[i].hashCode() == o.hashCode()){
```
Comment (r-zaiarnyi) [inline src/MyArrayList.java]: why do you need this `&& t[i].hashCode() == o.hashCode()` ?

File: src/MyArrayList.java [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    @Override
+    public void clear() {
+        t = (T[]) new Object[capacity];
+        size = 0;
+    }
+
+    @Override
+    public Object set(int index, Object element) {
```
Comment (r-zaiarnyi) [inline src/MyArrayList.java:121]: why does this method return object it set?

File: src/MyArrayList.java [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+        if (index >= size || index < 0) throw new IllegalArgumentException("Неправильно вказаний індекс");
+        t[index] = (T) element;
+        return element;
+    }
+
+    @Override
+    public int indexOf(Object o) {
+        for (int i = 0; i < size; i++){
+            if (t[i].equals(o) && t[i].hashCode() == o.hashCode()){
```
Comment (r-zaiarnyi) [inline src/MyArrayList.java]: why do you need this ? `&& t[i].hashCode() == o.hashCode()`

File: src/MyArrayList.java [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+
+
+
+
+
+    //////////// Not Implemented ////////////
+
+    @Override
+    public Iterator iterator() {
+        throw new NotImplementedException();
```
Comment (r-zaiarnyi) [inline src/MyArrayList.java]: please implement this to be able to traverse list in:
`for (T e : list) {
}`

File: src/MyArrayList.java [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    @Override
+    public Object[] toArray() {
+        throw new NotImplementedException();
```
Comment (r-zaiarnyi) [inline src/MyArrayList.java]: this is easy to implement, just return your array

File: src/MyArrayList.java [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+    }
+
+    @Override
+    public boolean remove(Object o) {
+        throw new NotImplementedException();
```
Comment (r-zaiarnyi) [inline src/MyArrayList.java]: please implement this as well

File: src/MyArrayListTest.java [hunk 11]
```diff
@@ -0,0 +1,111 @@
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
+
+class MyArrayListTest {
+
+    MyArrayList<Integer> myArrayList = new MyArrayList<>();
+
+    @Test
+    void checkAddAndConstructor(){
```
Comment (r-zaiarnyi) [inline src/MyArrayListTest.java:13]: where is a check in the test?

File: src/MyArrayListTest.java [hunk 12]
```diff
@@ -0,0 +1,111 @@
+import org.junit.jupiter.api.Test;
+import static org.junit.jupiter.api.Assertions.*;
+
+class MyArrayListTest {
+
+    MyArrayList<Integer> myArrayList = new MyArrayList<>();
```
Comment (r-zaiarnyi) [inline src/MyArrayListTest.java]: you should use:
`List<Integer> myArrayList = new MyArrayList<>();` instead

File: src/MyArrayListTest.java [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+        myArrayList.add(50);
+        assertEquals(5, myArrayList.size());
+        assertEquals(10, myArrayList.remove(0));
+        assertEquals(40, myArrayList.remove(2));
+        assertEquals(30, myArrayList.remove(1));
+        assertEquals(2, myArrayList.size());
+    }
+
+    @Test
+    void contains() {
```
Comment (r-zaiarnyi) [inline src/MyArrayListTest.java:87]: where is a code?

File: src/MyArrayListTest.java [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+        myArrayList.add(3);
+        myArrayList.add(4);
+        myArrayList.clear();
+        assertTrue(myArrayList.isEmpty(), "fun clear() doesn`t work");
+    }
+
+    @Test
+    void set() {
+        myArrayList.add(0, 0);
+        assertEquals(5, myArrayList.set(0, 5));
```
Comment (r-zaiarnyi) [inline src/MyArrayListTest.java]: what do you check here?

File: src/MyArrayListTest.java [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+        assertEquals(5, myArrayList.get(0));
+        assertEquals(1, myArrayList.size());
+    }
+
+    @Test
+    void indexOf() {
```
Comment (r-zaiarnyi) [inline src/MyArrayListTest.java:117]: please add more checks. for example when you remove items from front, does it shift all elements? indexes in that case has to be different.

File: src/MyLinkedList.java [hunk 16]
```diff
@@ -0,0 +1,258 @@
...
+    private int size = 0;
+
+    private class Node<E>{
+        private E val;
+        private Node next;
+        private Node prev;
+
+        private Node(E val) {
+            this.val = val;
+            size++;
```
Comment (r-zaiarnyi) [inline src/MyLinkedList.java]: what is this? why it is here?

File: src/MyLinkedList.java [hunk 17]
```diff
@@ -0,0 +1,258 @@
...
+        }
+
+    }
+    @Override
+    public boolean add(Object o) {
```
Comment (r-zaiarnyi) [inline src/MyLinkedList.java:31]: this method should increase size

File: src/MyLinkedList.java [hunk 18]
```diff
@@ -0,0 +1,258 @@
...
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        return true;
+    }
+
+    @Override
+    public void add(int index, Object element) {
```
Comment (r-zaiarnyi) [inline src/MyLinkedList.java:47]: this has to take E, not Object

File: src/MyLinkedList.java (hunk 17 above)
Comment (r-zaiarnyi) [inline src/MyLinkedList.java:31]: should be E instead of Object

File: src/MyLinkedList.java [hunk 19]
```diff
@@ -0,0 +1,258 @@
...
+        } else {
+            current.next.prev = current.prev;
+            current.prev.next = current.next;
+        }
+        size--;
+        return true;
+    }
+
+    @Override
+    public boolean contains(Object o) {
```
Comment (r-zaiarnyi) [inline src/MyLinkedList.java:127]: should use E

File: src/MyLinkedListTest.java [hunk 20]
```diff
@@ -0,0 +1,157 @@
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+
+import static org.junit.jupiter.api.Assertions.*;
+
+class MyLinkedListTest {
```
Comment (r-zaiarnyi) [inline src/MyLinkedListTest.java:6]: this test should be the same as test of ArrayList

File: src/MyStackTest.java [hunk 21]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Test
+    void nonEmpty() {
+        // check fun when we add some element(must be false)
+        myStack.push(1);
+        assertFalse(myStack.isEmpty(), "Stack is Empty!");
+    }
+
+    @Test
+    void isEmpty() {
```
Comment (r-zaiarnyi) [inline src/MyStackTest.java:17]: also add some test which check if `isEmpty() == false` after added elements and true after removal

**Review by r-zaiarnyi — CHANGES_REQUESTED** [review]
Please fix issues and send for review again

File: src/MyArrayList.java (hunk 1 above)
Comment (mkelshapik) [reply src/MyArrayList.java:34]: corrected

File: src/MyArrayList.java (hunk 2 above)
Comment (mkelshapik) [reply src/MyArrayList.java:53]: corrected

File: src/MyArrayList.java (hunk 3 above)
Comment (mkelshapik) [reply src/MyArrayList.java]: corrected

File: src/MyArrayList.java (hunk 4 above)
Comment (mkelshapik) [reply src/MyArrayList.java]: corrected

File: src/MyArrayList.java (hunk 5 above)
Comment (mkelshapik) [reply src/MyArrayList.java]: corrected

File: src/MyArrayList.java (hunk 7 above)
Comment (mkelshapik) [reply src/MyArrayList.java]: corrected

File: src/MyArrayList.java (hunk 6 above)
Comment (mkelshapik) [reply src/MyArrayList.java:121]: corrected

File: src/MyArrayList.java (hunk 9 above)
Comment (mkelshapik) [reply src/MyArrayList.java]: ready

File: src/MyArrayList.java (hunk 8 above)
Comment (mkelshapik) [reply src/MyArrayList.java]: Check

File: src/MyArrayList.java (hunk 10 above)
Comment (mkelshapik) [reply src/MyArrayList.java]: Check

File: src/MyArrayListTest.java (hunk 11 above)
Comment (mkelshapik) [reply src/MyArrayListTest.java:13]: Check

File: src/MyArrayListTest.java (hunk 12 above)
Comment (mkelshapik) [reply src/MyArrayListTest.java]: Check

File: src/MyArrayListTest.java (hunk 13 above)
Comment (mkelshapik) [reply src/MyArrayListTest.java:87]: Check

File: src/MyArrayListTest.java (hunk 14 above)
Comment (mkelshapik) [reply src/MyArrayListTest.java]: Check

File: src/MyArrayListTest.java (hunk 15 above)
Comment (mkelshapik) [reply src/MyArrayListTest.java:117]: Check

File: src/MyLinkedList.java (hunk 16 above)
Comment (mkelshapik) [reply src/MyLinkedList.java]: Corrected

File: src/MyLinkedList.java (hunk 18 above)
Comment (mkelshapik) [reply src/MyLinkedList.java:47]: annotation does not allow me do it

File: src/MyLinkedList.java (hunk 17 above)
Comment (mkelshapik) [reply src/MyLinkedList.java:31]: the same above

File: src/MyLinkedList.java (hunk 19 above)
Comment (mkelshapik) [reply src/MyLinkedList.java:127]: ))

**Pull request merged**